Source/AuthenticationInterceptor.swift

```diff
@@ -0,0 +1,402 @@
+            return
+        }
+
+        // Retry the request if the `Authenticator` verifies it was authenticated with a previous credential.
+        guard authenticator.isRequest(urlRequest, authenticatedWith: credential) else {
+            completion(.retry)
+            return
+        }
+
+        $mutableState.write { mutableState in
+            mutableState.requestsToRetry.append(completion)
+
+            guard !mutableState.isRefreshing else { return }
+
+            refresh(credential, for: session, insideLock: &mutableState)
+        }
+    }
+
+    // MARK: Refresh
+
+    private func refresh(_ credential: Credential, for session: Session, insideLock mutableState: inout MutableState) {
+        guard !isRefreshExcessive(insideLock: &mutableState) else {
+            let error = AuthenticationError.excessiveRefresh
+            handleRefreshFailure(error, insideLock: &mutableState)
+            return
+        }
+
+        mutableState.refreshTimestamps.append(ProcessInfo.processInfo.systemUptime)
+        mutableState.isRefreshing = true
+
+        authenticator.refresh(credential, for: session) { result in
+            self.$mutableState.write { mutableState in
+                switch result {
+                case let .success(credential):
+                    self.handleRefreshSuccess(credential, insideLock: &mutableState)
+
+                case let .failure(error):
+                    self.handleRefreshFailure(error, insideLock: &mutableState)
+                }
+            }
+        }
+    }
+
+    private func isRefreshExcessive(insideLock mutableState: inout MutableState) -> Bool {
+        guard let refreshWindow = mutableState.refreshWindow else { return false }
+
+        let refreshWindowMin = ProcessInfo.processInfo.systemUptime - refreshWindow.interval
+
+        let refreshAttemptsWithinWindow = mutableState.refreshTimestamps.reduce(into: 0) { attempts, refreshTimestamp in
+            guard refreshWindowMin <= refreshTimestamp else { return }
+            attempts += 1
+        }
+
+        let isRefreshExcessive = refreshAttemptsWithinWindow >= refreshWindow.maximumAttempts
+
+        return isRefreshExcessive
+    }
+
+    private func handleRefreshSuccess(_ credential: Credential, insideLock mutableState: inout MutableState) {
+        mutableState.credential = credential
+
+        let adaptOperations = mutableState.adaptOperations
+        let requestsToRetry = mutableState.requestsToRetry
+
+        mutableState.adaptOperations.removeAll()
+        mutableState.requestsToRetry.removeAll()
+
+        mutableState.isRefreshing = false
+
+        // Dispatch to queue to hop out of the mutable state lock
+        queue.async {
+            adaptOperations.forEach { self.adapt($0.urlRequest, for: $0.session, completion: $0.completion) }
+            requestsToRetry.forEach { $0(.retry) }
+        }
+    }
+
+    private func handleRefreshFailure(_ error: Error, insideLock mutableState: inout MutableState) {
```

Comment:
Any reason why we are not setting `mutableState.credential = nil` here?

We are running into an issue in cases where the refresh fails. We call the completion in refresh function with failure and then our enqueued API calls are retried with the older credential. Reason for that is AuthenticationInterceptor holds the older credential in its mutable state.